Flatten /prim and finish flatten

rok-cesnovar opened the pull request:
## Summary

Final PR in the flatten of `arr, mat, scal`. This PR flattens the rest of `/prim` and removes the checks that checked for arr, scal, mat include order.

## Tests

/

## Checklist

- [x] Math issue Closes #937 

- [x] Copyright holder: Rok Češnovar

By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

mcol commented:
What's the timeline for this? Is it waiting for the freeze first or do we need to push it before it?

rok-cesnovar commented:
We havent really discussed it. The freeze will start on the 11th (not sure what time, we usually take midnight anywhere on earth- https://www.worldtimeserver.com/time-zones/aoe/ or smth like that). 

But I am not sure whether to merge this during the freeze or not. Its technically not a bugfix but its not a new feature either. It is a huge PR and once again the review will take some time. 

Personally I do want to get this in before the next release so that we release the library in a clean state. I dont wont to force it tho.

mcol commented:
I would also like to see it make it, but given the size and impact on other PRs it would be more convenient to know that it could go in just after the freeze. @seantalts would that be possible?

In any case, I'm starting reviewing it now, let's see how far we can get.

rok-cesnovar commented:
Thanks Marco!

mcol commented on stan/math/prim/fun/mdivide_right_ldlt.hpp:
```diff
@@ -1,10 +1,10 @@
+#include <stan/math/prim/fun/LDLT_factor.hpp>
+#include <stan/math/prim/fun/mdivide_left_ldlt.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
 #include <stan/math/prim/err.hpp>
```
Move the err header first.

mcol commented on stan/math/prim/fun/mdivide_right_spd.hpp:
```diff
@@ -1,9 +1,9 @@
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/mdivide_left_spd.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
 #include <stan/math/prim/err.hpp>
```
err header first.

mcol commented on stan/math/prim/fun/mdivide_right_tri.hpp:
```diff
@@ -1,10 +1,10 @@
+#include <stan/math/prim/fun/mdivide_left_tri.hpp>
+#include <stan/math/prim/fun/promote_common.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
 #include <stan/math/prim/err.hpp>
```
err header first.

rok-cesnovar commented on stan/math/prim/fun/erfc.hpp:
```diff
@@ -0,0 +1,62 @@
+#define STAN_MATH_PRIM_FUN_ERFC_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/fun/erfc.hpp>
```
Thanks!

rok-cesnovar commented:
I addressed all comments in the first batch. 

I would definitely wait for at least https://github.com/stan-dev/math/pull/1558 to get merged before merging this one. The rest are currently either not ready yet or I would consider as bugfix PRs.

mcol commented on stan/math/prim/prob/wiener_lpdf.hpp:
```diff
@@ -31,11 +31,11 @@
+#include <stan/math/prim/fun/square.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
 #include <stan/math/prim/err.hpp>
-#include <stan/math/prim/scal/fun/size_zero.hpp>
```
Put err.hpp header first.

mcol commented on stan/math/rev/core/operator_unary_plus.hpp:
```diff
@@ -2,9 +2,9 @@
+#include <stan/math/prim/fun/is_nan.hpp>
 #include <stan/math/rev/core/precomp_v_vari.hpp>
-#include <stan/math/prim/scal/fun/constants.hpp>
+#include <stan/math/prim/fun/constants.hpp>
```
prim headers together.

mcol commented on stan/math/rev/fun/Eigen_NumTraits.hpp:
```diff
@@ -2,7 +2,7 @@
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/std_numeric_limits.hpp>
```
Put the rev headers together.

bob-carpenter commented:
#1525 is another big one waiting to be merged but that's probably not going to be ready to go for next PR (isn't code freeze today the 11th?).

mcol commented on stan/math/rev/fun/acosh.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/acosh.hpp>
+#include <stan/math/prim/fun/acosh.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/asinh.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/asinh.hpp>
+#include <stan/math/prim/fun/asinh.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/atanh.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/atanh.hpp>
+#include <stan/math/prim/fun/atanh.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/cbrt.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/cbrt.hpp>
+#include <stan/math/prim/fun/cbrt.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/determinant.hpp:
```diff
@@ -3,7 +3,7 @@
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/erfc.hpp:
```diff
@@ -2,8 +2,8 @@
-#include <stan/math/prim/scal/fun/erfc.hpp>
+#include <stan/math/prim/fun/constants.hpp>
+#include <stan/math/prim/fun/erfc.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/expm1.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/expm1.hpp>
+#include <stan/math/prim/fun/expm1.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/fdim.hpp:
```diff
@@ -2,8 +2,8 @@
-#include <stan/math/prim/scal/fun/is_any_nan.hpp>
+#include <stan/math/prim/fun/constants.hpp>
+#include <stan/math/prim/fun/is_any_nan.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/fmin.hpp:
```diff
@@ -3,9 +3,9 @@
+#include <stan/math/prim/fun/constants.hpp>
 #include <stan/math/rev/fun/is_nan.hpp>
-#include <stan/math/prim/scal/fun/is_nan.hpp>
+#include <stan/math/prim/fun/is_nan.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/grad.hpp:
```diff
@@ -2,7 +2,7 @@
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/fun/Eigen_NumTraits.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/grad_inc_beta.hpp:
```diff
@@ -3,9 +3,9 @@
+#include <stan/math/prim/fun/grad_2F1.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
 #include <stan/math/rev/fun/exp.hpp>
 #include <stan/math/rev/fun/fabs.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/inc_beta.hpp:
```diff
@@ -3,10 +3,10 @@
-#include <stan/math/prim/scal/fun/beta.hpp>
-#include <stan/math/prim/scal/fun/digamma.hpp>
+#include <stan/math/prim/fun/beta.hpp>
+#include <stan/math/prim/fun/digamma.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/inverse.hpp:
```diff
@@ -1,12 +1,12 @@
 #include <stan/math/prim/err.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/log1m_inv_logit.hpp:
```diff
@@ -2,8 +2,8 @@
+#include <stan/math/prim/fun/log1m_inv_logit.hpp>
+#include <stan/math/prim/fun/inv_logit.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/precomp_v_vari.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/log1p_exp.hpp:
```diff
@@ -3,7 +3,7 @@
 #include <stan/math/rev/core.hpp>
-#include <stan/math/prim/scal/fun/log1p_exp.hpp>
+#include <stan/math/prim/fun/log1p_exp.hpp>
 #include <stan/math/rev/fun/calculate_chain.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/log_determinant.hpp:
```diff
@@ -3,8 +3,8 @@
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/log_inv_logit.hpp:
```diff
@@ -2,8 +2,8 @@
+#include <stan/math/prim/fun/inv_logit.hpp>
+#include <stan/math/prim/fun/log_inv_logit.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/precomp_v_vari.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/log_mix.hpp:
```diff
@@ -2,8 +2,8 @@
-#include <stan/math/prim/scal/fun/log_mix.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <stan/math/prim/fun/log_mix.hpp>
 #include <stan/math/rev/fun/value_of.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/log_rising_factorial.hpp:
```diff
@@ -3,8 +3,8 @@
-#include <stan/math/prim/scal/fun/digamma.hpp>
+#include <stan/math/prim/fun/log_rising_factorial.hpp>
+#include <stan/math/prim/fun/digamma.hpp>
 #include <stan/math/rev/fun/digamma.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/log_softmax.hpp:
```diff
@@ -2,11 +2,11 @@
 #include <stan/math/rev/fun/typedefs.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/logit.hpp:
```diff
@@ -2,7 +2,7 @@
-#include <stan/math/prim/scal/fun/logit.hpp>
+#include <stan/math/prim/fun/logit.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/precomp_v_vari.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/matrix_power.hpp:
```diff
@@ -5,8 +5,8 @@
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/mdivide_left.hpp:
```diff
@@ -3,8 +3,8 @@
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/multiply_lower_tri_self_transpose.hpp:
```diff
@@ -2,7 +2,7 @@
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/ordered_constrain.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/functor/adj_jac_apply.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/positive_ordered_constrain.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/functor/adj_jac_apply.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/quad_form.hpp:
```diff
@@ -3,10 +3,10 @@
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <stan/math/prim/fun/quad_form.hpp>
 #include <stan/math/prim/err.hpp>
```
Move prim/err.hpp up.

mcol commented on stan/math/rev/fun/rows_dot_product.hpp:
```diff
@@ -3,7 +3,7 @@
 #include <stan/math/prim/err.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/simplex_constrain.hpp:
```diff
@@ -2,8 +2,8 @@
-#include <stan/math/prim/scal/fun/inv_logit.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/inv_logit.hpp>
 #include <stan/math/rev/functor/adj_jac_apply.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/softmax.hpp:
```diff
@@ -2,9 +2,9 @@
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/softmax.hpp>
 #include <stan/math/rev/functor/adj_jac_apply.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/tgamma.hpp:
```diff
@@ -2,8 +2,8 @@
-#include <stan/math/prim/scal/fun/tgamma.hpp>
+#include <stan/math/prim/fun/digamma.hpp>
+#include <stan/math/prim/fun/tgamma.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/fun/value_of_rec.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/value_of_rec.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/functor/coupled_ode_system.hpp:
```diff
@@ -3,7 +3,7 @@
+#include <stan/math/prim/fun/value_of.hpp>
 #include <stan/math/prim/err.hpp>
 #include <stan/math/rev/fun/value_of_rec.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/functor/gradient.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/functor/jacobian.hpp:
```diff
@@ -2,7 +2,7 @@
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
```
rev headers together.

mcol commented on stan/math/rev/functor/map_rect_reduce.hpp:
```diff
@@ -3,7 +3,7 @@
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
 #include <stan/math/rev/core/var.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
```
rev headers together.

rok-cesnovar commented:
> #1525 is another big one waiting to be merged but that's probably not going to be ready to go for next 

I was assuming it wont be ready, but yeah I am happ to wait for that PR too, if needed.

> (isn't code freeze today the 11th?).

Yes, 11th anywhere on the earth. If this gets approved today I will merge it once the deadline passes.

seantalts commented:
Agreed, thanks Rok! and thanks @mcol for stepping up to review :)

rok-cesnovar commented:
Thanks Marco for all the reviews. You really made this process as smooth as it could have been.
